src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettingsTemplatePackageProvider.cs

```diff
@@ -305,13 +312,60 @@ private async Task<InstallResult> InstallAsync(List<TemplatePackageData> package
                 throw new InvalidOperationException($"{nameof(installResult.TemplatePackage)} cannot be null when {nameof(installResult.Success)} is 'true'");
             }
 
+            // We check this after installation to make sure that we captured all templates inside a package
+            var (isUniqueIdentifier, errorMessage) = await EnsureUniqueIdentifierAsync(installResult, installRequest.Force, cancellationToken).ConfigureAwait(false);
+
+            if (!isUniqueIdentifier)
+            {
+                await installer.UninstallAsync(installResult.TemplatePackage, this, cancellationToken).ConfigureAwait(false);
+                return InstallResult.CreateFailure(installRequest, InstallerErrorCode.DuplicatedIdentity, errorMessage);
+            }
+
             lock (packages)
             {
                 packages.Add(((ISerializableInstaller)installer).Serialize(installResult.TemplatePackage));
             }
             return installResult;
         }
 
+        private async Task<(bool, string)> EnsureUniqueIdentifierAsync(InstallResult installResult, bool force, CancellationToken cancellationToken = default)
+        {
+            if (installResult.TemplatePackage is null)
+            {
+                throw new InvalidOperationException($"No package found in the result of this install operation");
+            }
+
+            using var scanResult = _scanner.Scan(installResult.TemplatePackage.MountPointUri, scanForComponents: false);
+            var installedTemplates = await _templatePackageManager.Value.GetTemplatesAsync(cancellationToken).ConfigureAwait(false);
+
+            foreach (ITemplate template in scanResult.Templates)
+            {
+                if (installedTemplates.FirstOrDefault(s =>
+                    s.Identity == template.Identity && !s.MountPointUri.Equals(template.MountPointUri, StringComparison.OrdinalIgnoreCase))
```

Comment:
```suggestion
                    s.Identity == template.Identity)
```

templates are not in cache yet, so s.MountPointUri.Equals(template.MountPointUri, StringComparison.OrdinalIgnoreCase) is never same.